### ISSUE TITLE
Early Stopping based on objective rather than gradient: Issue for the optimParallel package in R

### ISSUE THREAD

**Issue by shapenaji** [florafauna/optimParallel-python#2]
Hello! as stated above, this is really an issue for your R-package at
https://git.math.uzh.ch/florian.gerber/optimParallel/-/tree/master

(But I'm not affiliated with uzh so I don't think I can post an issue there)

I love the package! 

I would like institute an early stopping condition for L-BFGS-B that's based on my objective and not based on the gradient. 

Once my objective crosses a threshold, I don't need my gradient to improve anymore. This is already good enough. Despite a lot of looking, I haven't been able to find a parameter which allows me to do this. (abstol support in L-BFGS-B basically)

I could implement this by adding my own gradient, and then setting the value to 0 when it reaches a certain threshold, but then I need to re-implement your excellent search!

Basically I want to do: 
```{r}
# Gradient function in FGgenerator
g <- function(par){
        # if abstol is set and stopping_condition is met, return a 0 gradient
        if(!is.null(abstol) && stopping_condition(par, abstol)) return(0)
        evalFG(par) 
        i_g <<- i_g+1
        return(grad)
    }
```

Is this possible? am I overlooking an easier way to accomplish this?

Thank you for your time!

**Comment by florafauna** [comment]
From `?optim`:

     ‘factr’ controls the convergence of the ‘"L-BFGS-B"’ method.
          Convergence occurs when the reduction in the objective is
          within this factor of the machine tolerance. Default is
          ‘1e7’, that is a tolerance of about ‘1e-8’.

     ‘pgtol’ helps control the convergence of the ‘"L-BFGS-B"’ method.
          It is a tolerance on the projected gradient in the current
          search direction. This defaults to zero, when the check is
          suppressed.

But what you want is to stop when fn() is below a certain threshold (instead of the reduction of fn(), which is controlled by `factr`)?

**Comment by shapenaji** [comment]
Yep that's right. At this point its still possible that the gradient might have a value over my threshold. But additional function evaluations are very costly.

What I see right now is that I get into a nice range pretty quickly. But I don't know anything generally about the gradient in this "nice range", so while I can tune factr to stop in this specific case, it won't help with a slightly different search.

I even have considered just throwing an error with my parameter values when my objective reaches a desired range, then catching it and getting the parameters that way. (It's an awful hack, but it does have the virtue of working on non-parallel optim as well) 

But before I do that, I figured it was better to drop a line here.

**Comment by florafauna** [comment]
hm. I don't think this is an issue of optimParallel. optimParallel is intended to behave similar as optim, and this is the case here. 

But one way to solve your problem could be to modify fn such that it returns the last value if below the threshold. I gave it a try: 

```
generateFn <- function(fn, t){
    last <- NULL
    function(...){
        if(!is.null(last) && last < t)
            return(last)
        last <<- fn(...)
        last
    }
}

id <- function(x) {
    print(x)
    x
}

library(optimParallel)
options(optimParallel.loginfo=TRUE)
cl <- makeForkCluster(2,outfile="")     # set the number of processor cores
setDefaultCluster(cl=cl) # set 'cl' as default cluster
clusterExport(cl, "id")
     
fn <- generateFn(id, 9690)
optim(par=10000, fn=fn, method="L-BFGS-B")
fn <- generateFn(id, 9690)
optimParallel(par=10000, fn=fn)

fn <- generateFn(id, 4000)
optimParallel(par=10000, fn=fn)
fn <- generateFn(id, 4000)
optim(par=10000, fn=fn, method="L-BFGS-B")
```
Seems to work with `optim` and `optimParallel`. However, I don't see why `optimParallel` does more evaluations and terminates with a different message. But should not be a big issues since evaluates after the threshold is met are very fast.

**Comment by shapenaji** [comment]
Thank you very much, this works great. So basically you store a global value when you find something good enough and then you only return it, which makes the gradient zero.

I assume the reason that parallel takes longer is that it has to store the global value in several parallel sessions? And so it may continue the search a little longer?

**Comment by florafauna** [comment]
yes, this might be the reason.

**Comment by florafauna** [comment]
This is not an `optimParallel-python` issue.
